Elixir v1.17.0 announcement (#1761)
elixir-lang/elixir-lang.github.com: _posts/2024-06-12-elixir-v1-17-0-released.markdown

---
layout: post
title: Elixir v1.17 released
author: Andrea Leopardi
category: Releases
excerpt: Elixir v1.17 released with the start of the type system and Erlang/OTP 27 support
---

Elixir v1.17 has just been released. 🎉

> Elixir v1.17 doesn't implement full static typing, but new set-theoretic type

Reviewer: ```suggestion
Elixir v1.17 doesn't implement full static typing, but the new set-theoretic type
```

> system is making its first steps into the language in this release. This new version also adds support [Erlang/OTP 27](https://www.erlang.org/downloads/27), the latest and greatest Erlang release. You'll also find a new calendar-related data type (`Duration`) and a `Date.shift/2` function.

Reviewer: ```suggestion
system is making its first steps into the language in this release. This new version also adds support for [Erlang/OTP 27](https://www.erlang.org/downloads/27), the latest and greatest Erlang release. You'll also find a new calendar-related data type (`Duration`) and a `Date.shift/2` function.
```

---
layout: post
title: Elixir v1.17 released
author: Andrea Leopardi
category: Releases
excerpt: Elixir v1.17 released with the start of the type system and Erlang/OTP 27 support
---

Elixir v1.17 has just been released. 🎉

> Elixir v1.17 doesn't implement full static typing, but new set-theoretic type
> system is making its first steps into the language in this release. This new version also adds support [Erlang/OTP 27](https://www.erlang.org/downloads/27), the latest and greatest Erlang release. You'll also find a new calendar-related data type (`Duration`) and a `Date.shift/2` function.

Reviewer 2: I wouldn't start by saying what we don't do. :D

```suggestion
This release introduces set-theoretic types into a handful of language constructs. While there are still [many steps ahead of us](https://elixir-lang.org/blog/2023/06/22/type-system-updates-research-dev/), this important milestone already brings benefits to developers in the form of new warnings for common mistakes. This new version also adds support for [Erlang/OTP 27](https://www.erlang.org/downloads/27), the latest and greatest Erlang release. You'll also find a new calendar-related data type (`Duration`) and a `Date.shift/2` function.
```

PS: it includes @sabiwara suggestion from above too.

Let's dive in.

## Warnings from gradual set-theoretic types

> This release introduces gradual set-theoretic types to infer types from patterns and guards and use them to type check programs, enabling the Elixir compiler to find faults and bugs in codebases without requiring changes to existing software. The underlying principles, theory, and roadmap of our work have been outlined in ["The Design Principles of the Elixir Type System" by Giuseppe Castagna, Guillaume Duboc, José Valim](https://arxiv.org/abs/2306.06391).

Reviewer 2: ```suggestion
This release introduces gradual set-theoretic types to infer types from patterns and use them to type check programs, enabling the Elixir compiler to find faults and bugs in codebases without requiring changes to existing software. The underlying principles, theory, and roadmap of our work have been outlined in ["The Design Principles of the Elixir Type System" by Giuseppe Castagna, Guillaume Duboc, José Valim](https://arxiv.org/abs/2306.06391).
```

Reviewer 2: We don't do guards yet. I will update the release notes too.

    such as `%{adress: ...} = user` (notice `address` vs `adress`).

  * Accessing a key on a map or a struct that does not have the given key, such
    as `user.adress`.

  * Updating a struct or a map that does not define the given key, such as
    `%{user | adress: ...}`.

  * Invoking a function on non-modules, such as `user.address()`.

  * Capturing a function on non-modules, such as `&user.address/0`.

  * Attempting to call an anonymous function without an actual function, such as
    `user.()`.

  * Performing structural comparisons between structs, such as `my_date <
    ~D[2010-04-17]`.

  * Performing structural comparisons between non-overlapping types, such as
    `integer >= string`.

  * Building and pattern matching on binaries without the relevant specifiers,
    such as `<<name>>` (this warns because by default it expects an integer, it
    should have been `<<name::binary>>` instead).

  * Attempting to rescue an undefined exception or a struct that is not an
    exception.

  * Accessing a field that is not defined in a rescued exception.

Here's an example of how one of these warnings looks like:

> ![Example of a warning when accessing a mispelled struct field](/images/contents/type-warning-on-struct-field.png)

Reviewer 2: Good call on the editor picture. However, if we are including an editor picture, I think it should show the error inline (otherwise we will have to ask questions about when the warning will show up in editors, but they should just work today!).

These new warnings will help Elixir developers find bugs earlier and give more
confidence when refactoring code, especially around maps and structs. While
Elixir already emitted some of these warnings in the past, those were discovered
using syntax analysis. The new warnings are more reliable, precise, and with
better error messages. Keep in mind, however, that the Elixir typechecker only
infers types from patterns within the same function at the moment. Analysis from
guards and across function boundaries will be added in future releases. For more
details, see our new [reference document on gradual set-theoretic
types](https://hexdocs.pm/elixir/main/gradual-set-theoretic-types.html).

The type system was made possible thanks to a partnership between
[CNRS](https://www.cnrs.fr/) and [Remote](https://remote.com/). The development

> work is currently sponsored by [Fresha](https://www.fresha.com/),
> [Starfish*](https://starfish.team/), and [Dashbit](https://dashbit.co/).

Reviewer 2: ```suggestion
work is currently sponsored by [Fresha](https://www.fresha.com/)
([they are hiring!](https://www.fresha.com/careers/openings?department=engineering)),
[Starfish*](https://starfish.team/), and [Dashbit](https://dashbit.co/).
```

months:

```elixir
iex> ~D[2016-01-31] |> Date.shift(month: 1) |> Date.shift(month: 1)
~D[2016-03-29]
```

Still, durations are essential for building intervals, recurring events, and
modelling scheduling complexities found in the world around us. For `DateTime`s,
Elixir will correctly deal with time zone changes (such as Daylight Saving
Time). However, provisions are also available in case you want to surface
conflicts, such as shifting to a wall clock that does not exist, because the
clock has been moved forward by one hour. See `DateTime.shift/2` for examples.

Finally, we added a new `Kernel.to_timeout/1` function, which helps developers
normalize durations and integers to a timeout used by many APIs—like `Process`,
`GenServer`, and more. For example, to send a message after one hour, you can
now write:

```elixir
Process.send_after(pid, :wake_up, to_timeout(hour: 1))
```

## Learn more

Here are other notable changes in this release:

  * There are new `Keyword.intersect/2,3` functions to mirror the equivalent in
    the `Map` module.

  * A new Mix profiler was added, `mix profile.tprof`, which lets you use the

>     new tprof profiler released with Erlang/OTP 27. This profiler leads to the

Reviewer: ```suggestion
    new [tprof](https://erlang.org/documentation/doc-15.0/lib/tools-4.0/doc/html/tprof.html)
    profiler released with Erlang/OTP 27. This profiler leads to the
```

Reviewer 2: ```suggestion
    new [tprof](https://www.erlang.org/doc/apps/tools/tprof.html)
    profiler released with Erlang/OTP 27. This profiler leads to the
```

Same as above from @sabiwara but using a shorter URL.

---
layout: post

> title: Elixir v1.17 released
> author: Andrea Leopardi
> category: Releases
> excerpt: Elixir v1.17 released with the start of the type system and Erlang/OTP 27 support

Reviewer 2: ```suggestion
title: "Elixir v1.17 released: set-theoretic types in patterns, calendar durations, and Erlang/OTP 27 support"
author: Andrea Leopardi
category: Releases
excerpt: "Elixir v1.17 released: set-theoretic types in patterns, calendar durations, and Erlang/OTP 27 support"
```

The title is what is used by many places automatically, so I tried to add more information to it.

Author: Yesssirrrrr
